cli/cook/subcommands/show.py

```diff
@@ -0,0 +1,338 @@
+    instance = instance_job_pair[0]
+    job = instance_job_pair[1]
+
+    left = [['Cluster', cluster_name],
+            ['Host', instance['hostname']],
+            ['Slave', instance['slave_id']],
+            ['Job', '%s (%s)' % (job['name'], job['uuid'])]]
+    if len(instance['ports']) > 0:
+        left.append(['Ports Allocated', format_list(instance['ports'])])
+
+    right = [['Run Time', format_instance_run_time(instance)],
+             ['Instance Status', format_instance_status(instance)],
+             ['Job Status', format_job_status(job)]]
+    if 'exit_code' in instance:
+        right.append(['Exit Code', instance['exit_code']])
+
+    left_table = tabulate(left, tablefmt='plain')
+    right_table = tabulate(right, tablefmt='plain')
+    instance_tables = juxtapose_text(left_table, right_table)
+    return '\n=== Job Instance: %s ===\n\n%s' % (instance['task_id'], instance_tables)
+
+
+def tabulate_group(cluster_name, group):
+    """Given a group, returns a string containing a table for the group fields"""
+    left = [['Cluster', cluster_name]]
+    if group['host_placement']['type'] == 'all':
+        left.append(['Host Placement', 'all hosts'])
+    else:
+        left.append(['Host Placement', format_dict(group['host_placement'])])
+    if group['straggler_handling']['type'] == 'none':
+        left.append(['Straggler Handling', 'none'])
+    else:
+        left.append(['Straggler Handling', format_dict(group['straggler_handling'])])
+
+    right = [['# Completed', group['completed']],
+             ['# Running', group['running']],
+             ['# Waiting', group['waiting']]]
+
+    num_jobs = len(group['jobs'])
+    jobs = 'Job group contains %s job%s:\n%s' % (num_jobs, '' if num_jobs == 1 else 's', '\n'.join(group['jobs']))
+
+    left_table = tabulate(left, tablefmt='plain')
+    right_table = tabulate(right, tablefmt='plain')
+    group_tables = juxtapose_text(left_table, right_table)
+    return '\n=== Job Group: %s (%s) ===\n\n%s\n\n%s' % (group['uuid'], group['name'], group_tables, jobs)
+
+
+def show_data(cluster_name, data, format_fn):
+    """Iterates over the data collection and formats and prints each element"""
+    count = len(data)
+    if count > 0:
+        tables = [format_fn(cluster_name, datum) for datum in data]
+        output = '\n\n'.join(tables)
+        print(output)
+        print()
+    return count
+
+
+def query_cluster(cluster, uuids, pred, timeout, interval, make_request_fn):
+    """
+    Queries the given cluster for the given uuids with
+    an optional predicaite, pred, that must be satisfied
+    """
+
+    def satisfy_pred():
+        resp_ = make_request_fn(cluster, uuids)
+        return pred(resp_.json())
+
+    try:
+        resp = make_request_fn(cluster, uuids)
+        if resp.status_code == requests.codes.ok:
+            entities = resp.json()
+            if pred and not pred(entities):
+                entities = wait_until(satisfy_pred, timeout, interval)
```

Comment:
Is there any status reporting to the user while they are waiting as to what entities have finished and which entities they're waiting for?

Reply:
Good idea. Adding some reporting:

```bash
$ uuid=$(uuidgen) ; cs wait -t 30 21a93725-52e3-4f77-83b3-995bde32fd83 $(cs -s submit -g $uuid ls) c23f4cb7-84c0-4022-b240-bb9ef2765b94 $uuid
Will wait up to 30 seconds.
Waiting for the following jobs: 21a93725-52e3-4f77-83b3-995bde32fd83, 2297b96f-82b6-4ce2-a8ea-d01ecb2fb255...done
Waiting for instances of the following jobs: 21a93725-52e3-4f77-83b3-995bde32fd83...done
Waiting for the following job groups: c0172830-5664-4940-aa4e-2548f604981d...done
Waited for a total of 5.05 seconds.
```